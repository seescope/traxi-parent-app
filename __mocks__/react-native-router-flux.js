import React, { PropTypes } from 'react';
import { View } from 'react-native';

export const Actions = {
  createKid: jest.fn(),
  walkthrough: jest.fn(),
  congratulations: jest.fn(),
  thankyou: jest.fn(),
  areYouReady: jest.fn(),
  intro: jest.fn(),
};

export const Scene = (props) => <View {...props} />;

export const Router = ({ children }) => (
  <View>{children}</View>
);

Router.propTypes = {
  children: PropTypes.array.isRequired,
};
