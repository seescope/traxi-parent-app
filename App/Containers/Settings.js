// @flow
import React from 'react';
import { connect } from 'react-redux';
import {
  ScrollView,
  Image,
  Text,
  View,
  StyleSheet,
  PixelRatio,
  TouchableOpacity,
} from 'react-native';
import { Actions } from 'react-native-router-flux';
import Icon from 'react-native-vector-icons/FontAwesome';

import STYLES from '../Constants/Styles';
import Bar from '../Components/Bar';
import { GREY, LIGHT_GREY, VERY_LIGHT_GREY, GOOD } from '../Constants/Colours';
import addAdditionalChild from '../AsyncActions/AddAdditionalChild';

import type { ParentState } from '../Reducers/Parent';
import type { KidsState, Kid } from '../Reducers/Kids';
import type { RootState } from '../Reducers';

type StateProps = {
  parent: ParentState,
  kids: KidsState
};

type Props = StateProps & {
  onPress: () => void
};

type DispatchProps = {
  dispatch: (any) => any
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: VERY_LIGHT_GREY,
  },
  innerContainer: {
    flex: 1,
    backgroundColor: VERY_LIGHT_GREY,
    paddingTop: 16,
  },
  cardBody: {
    margin: 10,
  },
  cardHeaderStyle: {
    fontFamily: 'Raleway-ExtraBold',
    padding: 8,
    color: GREY,
    fontSize: 17,
    paddingBottom: 0,
  },
  headerContainer: {
    height: 40,
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  headerUnderlineStyle: {
    marginTop: -8,
    marginHorizontal: 9,
    height: 1 / PixelRatio.get(),
    width: 60,
    backgroundColor: LIGHT_GREY,
  },
  kidImage: {
    height: 40,
    width: 40,
    marginRight: 32,
    borderRadius: 20,
  },
  kidRow: {
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: 16,
  },
  kidName: {
    fontFamily: 'Raleway-ExtraBold',
    fontSize: 15,
    color: GREY,
  },
  infoRowTitle: {
    fontFamily: 'Raleway-ExtraBold',
    fontSize: 15,
    color: GREY,
  },
  info: {
    fontFamily: 'Raleway-Regular',
    fontSize: 15,
    color: GREY,
  },
  infoRow: {
    marginBottom: 16,
  },
});

type InfoRowProps = {
  info: ?string,
  title: string
};

const InfoRow = ({ info, title }: InfoRowProps) => (
  <View style={styles.infoRow}>
    <Text style={styles.infoRowTitle}>{title}</Text>
    <Text style={styles.info}>{info}</Text>
  </View>
);

const KidRow = ({ name, avatarURL }: Kid) => (
  <View style={styles.kidRow}>
    <Image
      resizeMode="contain"
      style={styles.kidImage}
      source={{ uri: avatarURL }}
    />
    <Text style={styles.kidName}>{name}</Text>
  </View>
);

const Settings = ({ parent, kids, onPress }: Props) => (
  <View style={styles.container}>
    <Bar
      title="Settings"
      buttonIcon="chevron-left"
      onPress={() => Actions.pop()}
    />
    <ScrollView style={styles.innerContainer}>
      <View style={STYLES.CARD}>
        <View style={styles.headerContainer}>
          <Text style={styles.cardHeaderStyle}>
            Your Details
          </Text>
        </View>
        <View style={styles.headerUnderlineStyle} />
        <View style={styles.cardBody}>
          <InfoRow info={parent.name} title="Name:" />
          <InfoRow info={parent.email} title="Email:" />
        </View>
      </View>

      <View style={STYLES.CARD}>
        <View style={styles.headerContainer}>
          <Text style={styles.cardHeaderStyle}>
            Your Children
          </Text>
        </View>
        <View style={styles.headerUnderlineStyle} />
        <View style={styles.cardBody}>
          {Object.keys(kids).map(UUID => <KidRow key={UUID} {...kids[UUID]} />)}
          <TouchableOpacity style={styles.kidRow} onPress={() => onPress()}>
            <Icon
              color={GOOD}
              name="plus-circle"
              style={styles.kidImage}
              size={43}
            />
            <Text style={styles.kidName}>Add another child</Text>
          </TouchableOpacity>
        </View>
      </View>

    </ScrollView>
  </View>
);

const mapStateToProps = (
  { parentState, kidsState }: RootState
): StateProps => ({
  parent: parentState,
  kids: kidsState,
});

export const mergeProps = (
  { parent, kids }: StateProps,
  { dispatch }: DispatchProps
): Props => ({
  parent,
  kids,
  onPress: () => {
    if (parent.upgradedAt) {
      dispatch(addAdditionalChild());
      Actions.setName();
      return;
    }

    Actions.upgrade();
  },
});

// $FlowFixMe
export default connect(mapStateToProps, null, mergeProps)(Settings);
