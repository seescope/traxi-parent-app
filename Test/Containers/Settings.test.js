import React from 'react';
import renderer from 'react-test-renderer';
import { Actions } from 'react-native-router-flux';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
jest.mock('../../App/AsyncActions/AddAdditionalChild', () =>
  jest.fn(() => ({
    type: 'TEST_ADD_ADDITIONAL_CHILD',
  })));

import Settings, { mergeProps } from '../../App/Containers/Settings';

const mockStore = configureStore([thunk]);
const testStore = mockStore({
  parentState: {
    name: 'Name',
    email: 'Email',
    kids: ['abc-123'],
  },
  kidsState: {
    'abc-123': {
      name: 'John Bobson',
    },
  },
  setupState: {
    kidUUID: 'abc-123',
  },
});

const SettingsComponent = () => (
  <Provider store={testStore}>
    <Settings />
  </Provider>
);

it('renders the <Settings> component', () => {
  const tree = renderer.create(<SettingsComponent />).toJSON();
  expect(tree).toMatchSnapshot();
});

describe('onPress', () => {
  describe('When an account has already been upgraded', () => {
    test('Dispatches addAdditional child and navigates to DeviceSetup', () => {
      const MOCK_DISPATCH = jest.fn();
      const { onPress } = mergeProps(
        {
          parent: {
            upgradedAt: 'some time',
          },
        },
        { dispatch: MOCK_DISPATCH }
      );

      onPress();

      expect(MOCK_DISPATCH).toHaveBeenCalledWith({
        type: 'TEST_ADD_ADDITIONAL_CHILD',
      });
      expect(Actions.deviceSetup).toHaveBeenCalled();
    });
  });

  describe('When an account has not been upgraded', () => {
    test('Navigates to your Upgrade', () => {
      const MOCK_DISPATCH = jest.fn();
      const { onPress } = mergeProps(
        {
          parent: {
            upgradedAt: undefined,
          },
        },
        { dispatch: MOCK_DISPATCH }
      );

      onPress();

      expect(Actions.upgrade).toHaveBeenCalled();
    });
  });
});
