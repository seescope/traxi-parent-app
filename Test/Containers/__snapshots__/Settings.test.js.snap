// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the <Settings> component 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F0F2F7",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#448AFF",
        "flexDirection": "row",
        "height": 72,
        "paddingHorizontal": 16,
        "paddingTop": 16,
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      isTVSelectable={true}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#FFF",
              "fontSize": 25,
            },
            undefined,
            Object {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
          ]
        }
      >
        
      </Text>
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "backgroundColor": "transparent",
          "color": "#FFF",
          "flex": 1,
          "fontFamily": "Raleway-ExtraBold",
          "fontSize": 24,
          "marginRight": 25,
          "textAlign": "center",
        }
      }
    >
      Settings
    </Text>
  </View>
  <RCTScrollView
    style={
      Object {
        "backgroundColor": "#F0F2F7",
        "flex": 1,
        "paddingTop": 16,
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "backgroundColor": "#FFF",
            "borderRadius": 4,
            "elevation": 6,
            "margin": 8,
            "marginBottom": 24,
            "padding": 8,
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 3,
              "width": 3,
            },
            "shadowOpacity": 0.16,
            "shadowRadius": 3,
            "width": 734,
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "height": 40,
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#545454",
                "fontFamily": "Raleway-ExtraBold",
                "fontSize": 17,
                "padding": 8,
                "paddingBottom": 0,
              }
            }
          >
            Your Details
          </Text>
        </View>
        <View
          style={
            Object {
              "backgroundColor": "#A6A8AB",
              "height": 0.5,
              "marginHorizontal": 9,
              "marginTop": -8,
              "width": 60,
            }
          }
        />
        <View
          style={
            Object {
              "margin": 10,
            }
          }
        >
          <View
            style={
              Object {
                "marginBottom": 16,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#545454",
                  "fontFamily": "Raleway-ExtraBold",
                  "fontSize": 15,
                }
              }
            >
              Name:
            </Text>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#545454",
                  "fontFamily": "Raleway-Regular",
                  "fontSize": 15,
                }
              }
            >
              Name
            </Text>
          </View>
          <View
            style={
              Object {
                "marginBottom": 16,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#545454",
                  "fontFamily": "Raleway-ExtraBold",
                  "fontSize": 15,
                }
              }
            >
              Email:
            </Text>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#545454",
                  "fontFamily": "Raleway-Regular",
                  "fontSize": 15,
                }
              }
            >
              Email
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#FFF",
            "borderRadius": 4,
            "elevation": 6,
            "margin": 8,
            "marginBottom": 24,
            "padding": 8,
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 3,
              "width": 3,
            },
            "shadowOpacity": 0.16,
            "shadowRadius": 3,
            "width": 734,
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "height": 40,
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#545454",
                "fontFamily": "Raleway-ExtraBold",
                "fontSize": 17,
                "padding": 8,
                "paddingBottom": 0,
              }
            }
          >
            Your Children
          </Text>
        </View>
        <View
          style={
            Object {
              "backgroundColor": "#A6A8AB",
              "height": 0.5,
              "marginHorizontal": 9,
              "marginTop": -8,
              "width": 60,
            }
          }
        />
        <View
          style={
            Object {
              "margin": 10,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "marginBottom": 16,
              }
            }
          >
            <Image
              resizeMode="contain"
              source={
                Object {
                  "uri": undefined,
                }
              }
              style={
                Object {
                  "borderRadius": 20,
                  "height": 40,
                  "marginRight": 32,
                  "width": 40,
                }
              }
            />
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#545454",
                  "fontFamily": "Raleway-ExtraBold",
                  "fontSize": 15,
                }
              }
            >
              John Bobson
            </Text>
          </View>
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            hitSlop={undefined}
            isTVSelectable={true}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "marginBottom": 16,
                "opacity": 1,
              }
            }
            testID={undefined}
            tvParallaxProperties={undefined}
          >
            <Text
              accessible={true}
              allowFontScaling={false}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": "#4CAF50",
                    "fontSize": 43,
                  },
                  Object {
                    "borderRadius": 20,
                    "height": 40,
                    "marginRight": 32,
                    "width": 40,
                  },
                  Object {
                    "fontFamily": "FontAwesome",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                ]
              }
            >
              
            </Text>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#545454",
                  "fontFamily": "Raleway-ExtraBold",
                  "fontSize": 15,
                }
              }
            >
              Add another child
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
